pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-		propagateErrors: r.cfg.PropagateErrors,
+	notIngestedCounts, toIngest, tsByTargets := f.splitByTargets(in, rules)
+	defer f.pools.putTsByTargets(tsByTargets)
 
-		requests: r.requestsTotal,
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	var requestWg sync.WaitGroup
+	requestWg.Add(len(tsByTargets))
+	errCh := make(chan error, len(tsByTargets))
+	for endpoint, ts := range tsByTargets {
+		f.newRequest(ctx, endpoint, ts, &requestWg, errCh)
 	}
+
+	go func() {
+		requestWg.Wait()
+		close(errCh)
+	}()
+
+	return notIngestedCounts, toIngest, errCh
 }
 
-type request struct {
-	ctx    context.Context
-	client *http.Client
-	pools  *pools
+type tsWithSampleCount struct {
+	ts     []mimirpb.PreallocTimeseries
+	counts TimeseriesCounts
+}
 
-	tsByEndpoint map[string]*[]mimirpb.PreallocTimeseries
+type tsByTargets map[string]tsWithSampleCount
 
-	// The rules which define:
-	// - which metrics get forwarded
-	// - where the metrics get forwarded to
-	// - whether the forwarded metrics should also be ingested (sent to ingesters)
-	rules           validation.ForwardingRules
-	timeout         time.Duration
-	propagateErrors bool
+type TimeseriesCounts struct {
+	SampleCount   int
+	ExemplarCount int
+}
 
-	requests prometheus.Counter
-	errors   *prometheus.CounterVec
-	samples  prometheus.Counter
-	latency  prometheus.Histogram
+func (t *TimeseriesCounts) count(ts mimirpb.PreallocTimeseries) {
+	t.SampleCount += len(ts.TimeSeries.Samples)
+	t.ExemplarCount += len(ts.TimeSeries.Exemplars)
 }
 
-func (r *request) Add(sample mimirpb.PreallocTimeseries) bool {
-	metric, err := extract.UnsafeMetricNameFromLabelAdapters(sample.Labels)
-	if err != nil {
-		// The only possible error is due to no metric name being defined, in which case we won't forward the sample
-		// to anywhere but we should still send it to the Ingesters.
-		return true
-	}
+// forwardingEndpointAssigner returns a func which takes a timeseries and assigns it endpoints to which it should be forwarded.
+func (f *forwarder) splitByTargets(tsSlice []mimirpb.PreallocTimeseries, rules validation.ForwardingRules) (TimeseriesCounts, []mimirpb.PreallocTimeseries, tsByTargets) {
+	// notIngestedCounts keeps track of the number of samples and exemplars that we don't send to the ingesters,
+	// we need to count these in order to later update some of the distributor's metrics correctly.
+	var notIngestedCounts TimeseriesCounts
+
+	// tsSliceWriteIdx is the index in the toIngest slice where we are writing TimeSeries to.
+	tsSliceWriteIdx := 0
+
+	tsByTargets := f.pools.getTsByTargets()
+	for tsSliceReadIdx, ts := range tsSlice {
+		ingest := false
+		var forwardingTarget string
+
+		metric, err := extract.UnsafeMetricNameFromLabelAdapters(ts.Labels)
+		if err != nil {
+			// Can't check whether a timeseries should be forwarded if it has no metric name.
+			// Send it to the Ingesters and don't forward it.
+			ingest = true
```

Comment:
Whould we log this?

Reply:
Oh okay, I see that the error can _only_ be "no metric name label". IMO it would be clearer if that function returned a bool instead.